src/client/datascience/kernel-launcher/localKernelSpecFinderBase.ts

```diff
@@ -0,0 +1,196 @@
+import { ReadWrite, Resource } from '../../common/types';
+import { PythonEnvironment } from '../../pythonEnvironments/info';
+import { getInterpreterKernelSpecName, isKernelRegisteredByUs } from '../jupyter/kernels/helpers';
+import { JupyterKernelSpec } from '../jupyter/kernels/jupyterKernelSpec';
+import { KernelSpecConnectionMetadata, PythonKernelConnectionMetadata } from '../jupyter/kernels/types';
+import { IJupyterKernelSpec } from '../types';
+
+type KernelSpecFileWithContainingInterpreter = { interpreter?: PythonEnvironment; kernelSpecFile: string };
+
+@injectable()
+export abstract class LocalKernelSpecFinderBase {
+    private cache?: KernelSpecFileWithContainingInterpreter[];
+    // Store our results when listing all possible kernelspecs for a resource
+    private kernelSpecCache = new Map<
+        string,
+        Promise<(KernelSpecConnectionMetadata | PythonKernelConnectionMetadata)[]>
+    >();
+
+    // Store any json file that we have loaded from disk before
+    private pathToKernelSpec = new Map<string, Promise<IJupyterKernelSpec | undefined>>();
+
+    constructor(
+        @inject(IFileSystem) protected readonly fs: IFileSystem,
+        @inject(IWorkspaceService) protected readonly workspaceService: IWorkspaceService
+    ) {}
+
+    @traceDecorators.error('List kernels failed')
+    protected async listKernelsWithCache(
+        cacheKey: string,
+        finder: () => Promise<(KernelSpecConnectionMetadata | PythonKernelConnectionMetadata)[]>
+    ): Promise<(KernelSpecConnectionMetadata | PythonKernelConnectionMetadata)[]> {
+        // If we have already searched for this resource, then use that.
+        const promise = this.kernelSpecCache.get(cacheKey);
+        if (promise) {
+            return promise;
+        }
+
+        this.kernelSpecCache.set(
+            cacheKey,
+            finder().then((items) => {
+                const distinctKernelMetadata = new Map<
+                    string,
+                    KernelSpecConnectionMetadata | PythonKernelConnectionMetadata
+                >();
+                traceInfoIf(
+                    !!process.env.VSC_JUPYTER_LOG_KERNEL_OUTPUT,
+                    `Kernel specs for ${cacheKey?.toString() || 'undefined'} are \n ${JSON.stringify(
+                        items,
+                        undefined,
+                        4
+                    )}`
+                );
+                items.map((kernelSpec) => {
+                    // Check if we have already seen this.
+                    if (!distinctKernelMetadata.has(kernelSpec.id)) {
+                        distinctKernelMetadata.set(kernelSpec.id, kernelSpec);
+                    }
+                });
+
+                return Array.from(distinctKernelMetadata.values()).sort((a, b) => {
+                    if (a.kernelSpec?.display_name === b.kernelSpec?.display_name) {
+                        return 0;
+                    } else {
+                        return 1;
```

Comment:
Unsure what the intent of the sort here is, items stay in the same position if display names are equal and then b always moved down the sort regardless of contents?

Reply:
This was old code, let me fix that.